multiplexers: replace priority of byobu with tmux (#508)
canonical/ubuntu-server-documentation: docs/reference/other-tools/terminal-multiplexers.md

Create a named session for easy identification:

```bash
tmux new-session -s myproject
```

List all sessions:

```bash
tmux list-sessions
# or the shorthand:
tmux ls
```

Attach to a specific session by name:

```bash
tmux attach-session -t myproject
# or the shorthand:
tmux a -t myproject
```

To attach to a session and detach all other clients connected to it:

```bash
tmux attach-session -t myproject -d
```

This is useful when you want exclusive access to a session.

Attach to the most recent session:

```bash
tmux attach
```

Kill a session:

```bash
tmux kill-session -t myproject
```

### Window management

Windows in tmux are like tabs in a browser - each window is a separate workspace within a session. You can have multiple windows open and switch between them.

**Creating and navigating windows**:

- {kbd}`Ctrl+b` {kbd}`c` - Create a new window
- {kbd}`Ctrl+b` {kbd}`n` - Switch to next window
- {kbd}`Ctrl+b` {kbd}`p` - Switch to previous window
- {kbd}`Ctrl+b` {kbd}`0-9` - Switch to window number 0-9
- {kbd}`Ctrl+b` {kbd}`,` - Rename current window

> - {kbd}`Ctrl+b` {kbd}`w` - List all windows and select from an interactive menu

Reviewer: ```suggestion
- {kbd}`Ctrl+B` {kbd}`C` - Create a new window
- {kbd}`Ctrl+B` {kbd}`N` - Switch to next window
- {kbd}`Ctrl+B` {kbd}`P` - Switch to previous window
- {kbd}`Ctrl+B` {kbd}`0-9` - Switch to window number 0-9
- {kbd}`Ctrl+B` {kbd}`,` - Rename current window
- {kbd}`Ctrl+B` {kbd}`W` - List all windows and select from an interactive menu
```

Author: some (not all - yeah confusion) of these can do something different when upper case - I'll leave them lower case

:::{note}
When you exit the shell in a window (by typing `exit` or pressing {kbd}`Ctrl+d`), that window closes. If it was the last window in a session, the entire tmux session terminates.
:::

### Pane management

Within a window, you can split your terminal into multiple panes to view different shells simultaneously.

**Creating and navigating panes**:

- {kbd}`Ctrl+b` {kbd}`%` - Split pane vertically (side by side)
- {kbd}`Ctrl+b` {kbd}`"` - Split pane horizontally (top and bottom)
- {kbd}`Ctrl+b` {kbd}`o` - Switch to next pane
- {kbd}`Ctrl+b` {kbd}`arrow-key` - Navigate between panes using arrow keys
- {kbd}`Ctrl+b` {kbd}`x` - Close current pane (prompts for confirmation)

Closing a pane terminates the shell running in it. If you want to preserve the process, detach from the session instead of closing panes.

**Resizing panes**:

> Once you have multiple panes, you can resize them. The easiest method is to **hold** {kbd}`Ctrl+b` and while holding use the arrow keys {kbd}`←` {kbd}`↑` {kbd}`→` {kbd}`↓` to adjust the size of the current pane.

Reviewer: ```suggestion
Once you have multiple panes, you can resize them. The easiest method is to **hold** {kbd}`Ctrl+B` and while holding use the arrow keys {kbd}`←` {kbd}`↑` {kbd}`→` {kbd}`↓` to adjust the size of the current pane.
```

Keep {kbd}`Ctrl+b` held down while pressing arrow keys repeatedly to resize continuously and smoothly.

Alternatively, you can use explicit resize commands:

- {kbd}`Ctrl+b` {kbd}`:` then type `resize-pane -D 10` (Down by 10 lines)
- {kbd}`Ctrl+b` {kbd}`:` then type `resize-pane -U 10` (Up by 10 lines)
- {kbd}`Ctrl+b` {kbd}`:` then type `resize-pane -L 10` (Left by 10 columns)

> - {kbd}`Ctrl+b` {kbd}`:` then type `resize-pane -R 10` (Right by 10 columns)

Author: some (not all - yeah confusion) of these can do something different when upper case - I'll leave them lower case

Some interactions like resizing panes can be complex, but custom key bindings can be configured to make whatever an administrator's common actions are more comfortable.

**Scrollback mode**:

Scrollback allows you to review previous command output that has scrolled off the screen. tmux maintains a buffer (history) of terminal output.

> - {kbd}`Ctrl+b` {kbd}`[` - Enter scrollback mode

Reviewer: ```suggestion
- {kbd}`Ctrl+B` {kbd}`[` - Enter scrollback mode
```

> Once in scrollback mode, use arrow keys, {kbd}`Page Up`/{kbd}`Page Down`, or vi-style keys ({kbd}`j`/{kbd}`k`) to navigate. Press {kbd}`q` to exit scrollback mode. The default scrollback buffer stores 2000 lines, but this can be increased in the configuration (see the Configuration section).

Reviewer: ```suggestion
Once in scrollback mode, use arrow keys, {kbd}`Page Up`/{kbd}`Page Down`, or vi-style keys ({kbd}`J`/{kbd}`K`) to navigate. Press {kbd}`Q` to exit scrollback mode. The default scrollback buffer stores 2000 lines, but this can be increased in the configuration (see the Configuration section).
```

**Multi-user session sharing** (different users):

For different users to share a session, you need to create a shared socket:

User 1 creates a session with a shared socket:
```bash
tmux -S /tmp/shared-session new-session -s collaboration
chmod 777 /tmp/shared-session
```

User 2 connects using the shared socket:
```bash
tmux -S /tmp/shared-session attach-session -t collaboration
```

:::{note}
For multi-user sharing, ensure both users have appropriate permissions on the socket file. Consider security implications when sharing sessions between different users.
:::

### Configuration

tmux can be customized through the `~/.tmux.conf` file. The configuration possibilities are extensive and depend on your specific needs. Here are two commonly useful examples:

**Enable mouse support** for easier pane/window navigation and resizing:

```bash
# Enable mouse support
set -g mouse on
```

**Increase scrollback history** to review more command output:

```bash
# Increase scrollback buffer size from default 2000 to 10000 lines
set-option -g history-limit 10000
```

After creating or modifying `~/.tmux.conf`, reload it from within tmux:

```bash
tmux source-file ~/.tmux.conf
```

> Or use {kbd}`Ctrl+b` {kbd}`:` and type `source-file ~/.tmux.conf`.

Reviewer: ```suggestion
Or use {kbd}`Ctrl+B` {kbd}`:` and type `source-file ~/.tmux.conf`.
```

Reloading the configuration preserves all your existing sessions, windows, and panes. The alternative would be to exit and restart tmux, but that would lose your current work environment.

### Scripting and automation

tmux can be scripted to automate the creation of complex session layouts with multiple windows and panes. Scripts can use commands like `new-session`, `split-window`, `new-window`, and `send-keys` to set up your preferred environment automatically. Sessions are identified by name, windows by their index within a session, and panes by their index within a window. For detailed scripting syntax and options, refer to the {manpage}`tmux(1)` manual page.

### tmux plugins

The tmux ecosystem includes a plugin manager called TPM (Tmux Plugin Manager) that extends functionality. Install it from the Ubuntu archive:

```bash
sudo apt install tmux-plugin-manager
```

:::{note}

> While tmux plugins can be powerful, be aware that they introduce third-party code and configuration into your system. The same considerations discussed in the [third-party repository usage](https://documentation.ubuntu.com/server/explanation/software/third-party-repository-usage/) documentation apply here - evaluate trustworthiness, security implications, and maintenance status before using plugins, especially in production environments.

Reviewer: ```suggestion
While tmux plugins can be powerful, be aware that they introduce third-party code and configuration into your system. The same considerations discussed in the {ref}`third-party-repository-usage` documentation apply here - evaluate trustworthiness, security implications, and maintenance status before using plugins, especially in production environments.
```
